From 85db8100819737bd662c9530013c2b90e27db688 Mon Sep 17 00:00:00 2001
From: Thomas Bouvier <contact@thomas-bouvier.io>
Date: Tue, 17 Oct 2023 12:11:34 +0200
Subject: [PATCH 1/2] feat: compile using cpp17 from version pytorch 2.1

Signed-off-by: Thomas Bouvier <contact@thomas-bouvier.io>
---
 horovod/torch/CMakeLists.txt | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/horovod/torch/CMakeLists.txt b/horovod/torch/CMakeLists.txt
index 7234d741ef..d54329d9ff 100644
--- a/horovod/torch/CMakeLists.txt
+++ b/horovod/torch/CMakeLists.txt
@@ -63,7 +63,9 @@ list(APPEND PYTORCH_LINKER_LIBS ${Pytorch_LIBRARIES})
 parse_version(${Pytorch_VERSION} VERSION_DEC)
 add_definitions(-DPYTORCH_VERSION=${VERSION_DEC} -DTORCH_API_INCLUDE_EXTENSION_H=1)
 set(Pytorch_CXX11 ${Pytorch_CXX11} PARENT_SCOPE)
-if(NOT Pytorch_VERSION VERSION_LESS "1.5.0")
+if(NOT Pytorch_VERSION VERSION_LESS "2.1.0")
+    set(CMAKE_CXX_STANDARD 17)
+elseif(NOT Pytorch_VERSION VERSION_LESS "1.5.0")
     set(CMAKE_CXX_STANDARD 14)
 endif()
 

From 55e4500fb3f43f3dbb6764094d162ed1f85329f5 Mon Sep 17 00:00:00 2001
From: Enrico Minack <github@enrico.minack.dev>
Date: Sat, 30 Dec 2023 14:07:25 +0100
Subject: [PATCH 2/2] Build Pytorch (and Gloo) with C++17 for >= 2.1.0

Signed-off-by: Enrico Minack <github@enrico.minack.dev>
---
 CMakeLists.txt               |  4 ++--
 horovod/torch/CMakeLists.txt | 20 ++++++++++++++------
 2 files changed, 16 insertions(+), 8 deletions(-)

diff --git a/CMakeLists.txt b/CMakeLists.txt
index 2c12719026..e0649013b7 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -375,7 +375,7 @@ if(HAVE_GLOO AND ((DEFINED Tensorflow_CXX11 AND NOT Tensorflow_CXX11) OR (DEFINE
 endif()
 
 # Gloo for c++17 TF
-if(HAVE_GLOO AND (DEFINED Tensorflow_CXX17 AND Tensorflow_CXX17))
+if(HAVE_GLOO AND ((DEFINED Tensorflow_CXX17 AND Tensorflow_CXX17) OR (DEFINED Pytorch_CXX17 AND Pytorch_CXX17)))
     file(COPY ${PROJECT_SOURCE_DIR}/third_party/gloo/ DESTINATION ${PROJECT_SOURCE_DIR}/third_party/compatible17_gloo)
     file(READ ${PROJECT_SOURCE_DIR}/third_party/compatible17_gloo/gloo/CMakeLists.txt GLOO_CMAKE)
     string(REPLACE "gloo " "compatible17_gloo " GLOO_CMAKE "${GLOO_CMAKE}")
@@ -385,7 +385,7 @@ if(HAVE_GLOO AND (DEFINED Tensorflow_CXX17 AND Tensorflow_CXX17))
     string(PREPEND GLOO_CMAKE "set(CMAKE_CXX_STANDARD 17)\n")
     file(WRITE ${PROJECT_SOURCE_DIR}/third_party/compatible17_gloo/CMakeLists.txt "${GLOO_CMAKE}")
     add_subdirectory(third_party/compatible17_gloo)
-    if (Tensorflow_CXX11)
+    if (Tensorflow_CXX11 OR Pytorch_CXX11)
         target_compile_definitions(compatible17_gloo PRIVATE _GLIBCXX_USE_CXX11_ABI=1)
     else()
        target_compile_definitions(compatible17_gloo PRIVATE _GLIBCXX_USE_CXX11_ABI=0)
diff --git a/horovod/torch/CMakeLists.txt b/horovod/torch/CMakeLists.txt
index d54329d9ff..d94768770f 100644
--- a/horovod/torch/CMakeLists.txt
+++ b/horovod/torch/CMakeLists.txt
@@ -14,6 +14,16 @@ if(NOT PYTORCH_FOUND)
     return()
 endif()
 
+if(Pytorch_VERSION VERSION_GREATER_EQUAL "2.1")
+    set(CMAKE_CXX_STANDARD 17)
+    set(Pytorch_CXX17 TRUE)
+elseif(Pytorch_VERSION VERSION_GREATER_EQUAL "1.5")
+    set(CMAKE_CXX_STANDARD 14)
+    set(Pytorch_CXX17 FALSE)
+else()
+    set(Pytorch_CXX17 FALSE)
+endif()
+
 # Append version number into metadata
 file(APPEND "${CMAKE_LIBRARY_OUTPUT_DIRECTORY_ROOT}/metadata.json" "\"pytorch\": \"${Pytorch_VERSION}\",\n")
 
@@ -39,7 +49,9 @@ execute_process(COMMAND ${PY_EXE} -c "from distutils.sysconfig import get_python
 include_directories(SYSTEM ${PYTHON_INCLUDE_PATH})
 set(CMAKE_CXX_FLAGS "${Pytorch_COMPILE_FLAGS} ${CMAKE_CXX_FLAGS}")
 if(HAVE_GLOO)
-    if (Pytorch_CXX11)
+    if (Pytorch_CXX17)
+        list(APPEND PYTORCH_LINKER_LIBS compatible17_gloo)
+    elseif (Pytorch_CXX11)
         list(APPEND PYTORCH_LINKER_LIBS gloo)
     else()
         list(APPEND PYTORCH_LINKER_LIBS compatible_gloo)
@@ -63,11 +75,7 @@ list(APPEND PYTORCH_LINKER_LIBS ${Pytorch_LIBRARIES})
 parse_version(${Pytorch_VERSION} VERSION_DEC)
 add_definitions(-DPYTORCH_VERSION=${VERSION_DEC} -DTORCH_API_INCLUDE_EXTENSION_H=1)
 set(Pytorch_CXX11 ${Pytorch_CXX11} PARENT_SCOPE)
-if(NOT Pytorch_VERSION VERSION_LESS "2.1.0")
-    set(CMAKE_CXX_STANDARD 17)
-elseif(NOT Pytorch_VERSION VERSION_LESS "1.5.0")
-    set(CMAKE_CXX_STANDARD 14)
-endif()
+set(Pytorch_CXX17 ${Pytorch_CXX17} PARENT_SCOPE)
 
 # PyTorch SOURCES
 # Later versions of PyTorch that use ROCm's hipify step will rename files.
